src/mito2/src/cache/file_cache.rs

```diff
@@ -0,0 +1,397 @@
+use object_store::{ErrorKind, Metakey, ObjectStore, Reader};
+use snafu::ResultExt;
+use store_api::storage::RegionId;
+
+use crate::cache::FILE_TYPE;
+use crate::error::{OpenDalSnafu, Result};
+use crate::metrics::{CACHE_BYTES, CACHE_HIT, CACHE_MISS};
+use crate::sst::file::FileId;
+
+/// Subdirectory of cached files.
+const FILE_DIR: &str = "files";
+
+/// A file cache manages files on local store and evict files based
+/// on size.
+#[derive(Debug)]
+pub(crate) struct FileCache {
+    /// Local store to cache files.
+    local_store: ObjectStore,
+    /// Cached file directory under cache home.
+    file_dir: String,
+    /// Index to track cached files.
+    ///
+    /// File id is enough to identity a file uniquely.
+    memory_index: Cache<IndexKey, IndexValue>,
+}
+
+impl FileCache {
+    /// Creates a new file cache.
+    pub(crate) fn new(
+        local_store: ObjectStore,
+        cache_home: String,
+        capacity: ReadableSize,
+    ) -> FileCache {
+        // Stores files under `cache_home/{FILE_DIR}`.
+        let file_dir = cache_file_dir(&cache_home);
+        let cache_store = local_store.clone();
+        let cache_file_dir = file_dir.clone();
+        let memory_index = Cache::builder()
+            .weigher(|_key, value: &IndexValue| -> u32 {
+                // We only measure space on local store.
+                value.file_size
+            })
+            .max_capacity(capacity.as_bytes())
+            .async_eviction_listener(move |key, value, cause| {
+                let store = cache_store.clone();
+                let file_path = cache_file_path(&cache_file_dir, *key);
+                async move {
+                    if let RemovalCause::Replaced = cause {
+                        // The cache is replaced by another file. This is unexpected, we don't remove the same
+                        // file but updates the metrics as the file is already replaced by users.
+                        CACHE_BYTES.with_label_values(&[FILE_TYPE]).sub(value.file_size.into());
+                        warn!("Replace existing cache {} for region {} unexpectedly", file_path, key.0);
+                        return;
+                    }
+
+                    match store.delete(&file_path).await {
+                        Ok(()) => {
+                            CACHE_BYTES.with_label_values(&[FILE_TYPE]).sub(value.file_size.into());
+                        }
+                        Err(e) => {
+                            warn!(e; "Failed to delete cached file {} for region {}", file_path, key.0);
+                        }
+                    }
+                }
+                .boxed()
+            })
+            .build();
+        FileCache {
+            local_store,
+            file_dir,
+            memory_index,
+        }
+    }
+
+    /// Puts a file into the cache.
+    ///
+    /// Callers should ensure the file is in the correct path.
+    pub(crate) async fn put(&self, key: IndexKey, value: IndexValue) {
```

Comment:
Do callers need to write both local and remote, and then update the cache?

Reply:
Callers should update the cache after they write local. This cache doesn't care about whether the file is stored in remote.

Reply:
How can callers "ensure the file is in the correct path"? Assuming I'm the caller, I have no idea where to store the file that only `FileCache` knows.

Reply:
You're right. I added a method to get the path in my POC branch but not in this PR. I'll port it to this PR.

The file cache is a component of the write-through cache. It is for internal use only. Maybe I should change `Callers` to the `WriteCache`.